crates/iroha_core/src/smartcontracts/wasm/cache.rs

```diff
@@ -0,0 +1,79 @@
+use iroha_data_model::parameter::SmartContractParameters;
+use wasmtime::{Engine, Module, Store};
+
+use crate::{
+    prelude::WorldReadOnly,
+    smartcontracts::{
+        wasm,
+        wasm::{state::executor::ValidateTransaction, RuntimeFull},
+    },
+    state::StateTransaction,
+};
+
+/// Executor related things (linker initialization, module instantiation, memory free)
+/// takes significant amount of time in case of single peer transactions handling.
+/// (https://github.com/hyperledger/iroha/issues/3716#issuecomment-2348417005).
+/// So this cache is used to share `Store` and `Instance` for different transaction validation.
+pub struct WasmCache<'world, 'block, 'state> {
+    cache: Option<RuntimeFull<ValidateTransaction<'world, 'block, 'state>>>,
+}
+
+impl<'world, 'block, 'state> WasmCache<'world, 'block, 'state> {
+    /// Constructor
+    pub fn new() -> Self {
+        Self { cache: None }
+    }
+
+    /// Hack to pass borrow checker. Should be used only when there is no data in `Store`.
+    #[allow(unsafe_code)]
+    pub fn change_lifetime<'l>(wasm_cache: &'l mut WasmCache) -> &'l mut Self {
+        if let Some(cache) = wasm_cache.cache.as_ref() {
+            assert!(cache.store.data().is_none());
+        }
```

Comment:
```suggestion
        }
        /// SAFETY: since we have ensured that `cache.store.data()` is `None`, the lifetime parameters we are transmuting are not used by any references
```